3scaleAdapter/pkg/threescale/proxy_conf.go

```diff
@@ -0,0 +1,158 @@
+// cache and when it should expire, and limit - the total number of entries that can be
+// stored in the cache at any given time.
+func NewProxyConfigCache(cacheTTL time.Duration, limit int) *ProxyConfigCache {
+	pcc := &ProxyConfigCache{
+		limit: limit,
+		ttl:   cacheTTL,
+		cache: make(map[string]proxyStore, limit),
+	}
+	return pcc
+}
+
+// FlushCache iterates over the items in the cache and purges any expired items
+func (pc *ProxyConfigCache) FlushCache() {
+	forDeletion := pc.markForDeletion()
+	pc.mutex.Lock()
+	for _, k := range forDeletion {
+		delete(pc.cache, k)
+	}
+	pc.mutex.Unlock()
+}
+
+// StartFlushWorker starts a background process that periodically carries out the
+// functionality provided by FlushCache
+func (pc *ProxyConfigCache) StartFlushWorker() error {
+	if pc.workerRunning {
+		return errors.New("worker has already been started")
+	}
+
+	pc.stopWorker = make(chan bool)
+	go pc.flushCache(pc.stopWorker)
+	pc.workerRunning = true
+	return nil
+}
+
+// StopFlushWorker stops a background process started by StartFlushWorker if it has been started
+// Returns an error if the background task is not running
+func (pc *ProxyConfigCache) StopFlushWorker() error {
+	if pc.stopWorker == nil || !pc.workerRunning {
+		return errors.New("worker has not been started")
+	}
+
+	pc.stopWorker <- true
+	close(pc.stopWorker)
+	pc.workerRunning = false
+	return nil
+}
+
+func (pc *ProxyConfigCache) get(cfg *config.Params, c *sysC.ThreeScaleClient) (sysC.ProxyConfigElement, error) {
+	var conf sysC.ProxyConfigElement
+	var err error
+
+	cacheKey := pc.getCacheKeyFromCfg(cfg)
+
+	pc.mutex.RLock()
+	e, ok := pc.cache[cacheKey]
+	pc.mutex.RUnlock()
+
+	if !ok {
+		conf, err = getFromRemote(cfg, c)
+		if err == nil {
+			go pc.set(cfg, conf)
+		}
+	} else {
+		log.Debugf("proxy config fetched from cache for service id %s", cfg.ServiceId)
+		conf = e.element
+	}
+
+	return conf, err
+}
+
+func (pc *ProxyConfigCache) set(cfg *config.Params, e sysC.ProxyConfigElement) {
+	cacheKey := pc.getCacheKeyFromCfg(cfg)
+
+	pc.mutex.Lock()
+	defer pc.mutex.Unlock()
+	if len(pc.cache) < pc.limit {
+		pc.cache[cacheKey] = proxyStore{
+			expiresAt: time.Now().Add(pc.ttl),
+			element:   e,
+		}
+	}
+}
+
+func (pc *ProxyConfigCache) getCacheKeyFromCfg(cfg *config.Params) string {
+	return fmt.Sprintf(cacheKey, cfg.SystemUrl, cfg.ServiceId)
+}
+
+func (pc *ProxyConfigCache) flushCache(exitC chan bool) {
+	for {
+
+		select {
+		case stop := <-exitC:
+			if stop {
+				log.Debugf("stopping cache flushing worker")
+				return
+			}
+		default:
+			pc.FlushCache()
+			<-time.After(pc.ttl)
+		}
+
+	}
+}
+
+func (pc *ProxyConfigCache) markForDeletion() []string {
+	var forDeletion []string
+	now := time.Now()
+	pc.mutex.RLock()
+	defer pc.mutex.RUnlock()
+	for k, v := range pc.cache {
+		if now.After(v.expiresAt) {
+			log.Debugf("proxy config with key %s is being purged from cache", k)
+			forDeletion = append(forDeletion, k)
+		}
+	}
+	return forDeletion
```

Comment:
@unleashed I've ran some benchmarks on this and the previous implementation. Changing this to read lock showed a 3-4  times increase in performance, since the reads can still happen from the cache while we are figuring out what has expired